Makefile

```diff
@@ -186,18 +190,26 @@ build: setup go-build
 PILOT_GO_BINS:=${ISTIO_BIN}/pilot-discovery ${ISTIO_BIN}/pilot-agent \
                ${ISTIO_BIN}/istioctl ${ISTIO_BIN}/sidecar-initializer
 $(PILOT_GO_BINS): depend
-	bin/gobuild.sh ${ISTIO_BIN}/$(@F) istio.io/istio/pilot/tools/version ./pilot/cmd/$(@F)
+	bin/gobuild.sh $@ istio.io/istio/pilot/tools/version ./pilot/cmd/$(@F)
+
+# Non-static istioctls. These are typically a build artifact so placed in out/ rather than bin/ .
+# The if/findstring is used to replace osx with darwin, win with windows and anything else (currently linux) with the suffix after the dash.
+# Cross-compiling for Mac on a Linux system seems to have fewer issues if CGO_ENABLED=0 (this is now handled by gobuild.sh).
+ISTIOCTL_LIST:=istioctl-linux istioctl-osx istioctl-win.exe
+$(foreach TGT,$(ISTIOCTL_LIST),$(eval ${OUT}/$(TGT): depend; \
+        STATIC=0 GOOS=$(if $(findstring osx,$(TGT)),darwin,$(if $(findstring win,$(TGT)),windows,$(subst istioctl-,,$(TGT)))) \
+                bin/gobuild.sh $$@ istio.io/istio/pilot/tools/version ./pilot/cmd/istioctl ))
```

Comment:
I do appreciate the ability to handle multiple versions of target.. But in the interest of keeping the makefile dumb and simple, I would seriously suggest simply creating three makefile targets for istioctl-osx, istioctl-win, istioctl-linux, each with appropriate hardcoded string (e.g., GOOS=linux/osx/windows). Its much more grokkable.. Yes, there will be more lines in makefile, but its boiler plate dumb make targets.

The more advanced or complicated scripting we add to Makefile, the closer it gets to the earlier bazel state. Some things won't work, people start complaining, and so on..

Reply:
+1, you forgot "like bazel, but twice as slow".
Please keep it simple. A makefile approaching 1000 lines is incomprehensible.